### PR TITLE
Per Graphic Shaders

### PR DESCRIPTION
**Pull request by MattTuttle**
This allows all `Graphic` objects to have their own shader. I moved all shader classes to `haxepunk.graphics.shaders.*` and renamed the fullscreen shader to `SceneShader`.

OpenGL buffers are filled within the shader class now. This optimizes the data sent to OpenGL and allows someone to change the attributes in their shader. I think eventually the GLBuffers should be created in the Shader class but for now it is still in HardwareRenderer.

Also, I'm not sure I'm 100% happy with the idea of TextureShader and ColorShader being singletons. There is probably a better way to do that as well.

### CODE REVIEW

File: haxepunk/graphics/atlas/DrawCommand.hx [hunk 1]
```diff
@@ -12,24 +12,22 @@ private class RenderData
...
-	public var tx3(get, set):Float; inline function get_tx3() return _data[8]; inline function set_tx3(v:Float) return _data[8] = v;
-	public var ty3(get, set):Float; inline function get_ty3() return _data[9]; inline function set_ty3(v:Float) return _data[9] = v;
-	public var uvx3(get, set):Float; inline function get_uvx3() return _data[10]; inline function set_uvx3(v:Float) return _data[10] = v;
-	public var uvy3(get, set):Float; inline function get_uvy3() return _data[11]; inline function set_uvy3(v:Float) return _data[11] = v;
-	public var red(get, set):Float; inline function get_red() return _data[12]; inline function set_red(v:Float) return _data[12] = v;
-	public var green(get, set):Float; inline function get_green() return _data[13]; inline function set_green(v:Float) return _data[13] = v;
-	public var blue(get, set):Float; inline function get_blue() return _data[14]; inline function set_blue(v:Float) return _data[14] = v;
-	public var alpha(get, set):Float; inline function get_alpha() return _data[15]; inline function set_alpha(v:Float) return _data[15] = v;
+	public var tx1:Float;
```
Comment (bendmorris) [inline haxepunk/graphics/atlas/DrawCommand.hx]: These should probably be initialized with zeroes.

**Comment by bendmorris** [comment]
Do you have an example of how this would be used? Would I create my own Shader classes? SceneShader lets you build one from a text asset; should we provide a similar convenience method for graphic shaders?

Do you think we should also provide some builtin custom shaders within HaxePunk?

**Comment by MattTuttle** [comment]
As of now you have to subclass Shader/TextureShader/ColorShader. It needs some work to be easier to use. Having some example shaders is probably a good idea but I don't know if they need to be builtin.

File: haxepunk/graphics/atlas/DrawCommand.hx (hunk 1 above)
Comment (MattTuttle) [inline haxepunk/graphics/atlas/DrawCommand.hx]: Good point. I added an initial value.

**Comment by MattTuttle** [comment]
I decided to eat the slight performance hit and add checks in `Shader.prepare()`. The shader data is no longer interleaved which may be a problem for data locality so it may not be the smartest change.

The nice part is that if you set `position`, `texCoord`, or `color` in a Shader subclass it will automatically prepare the correct values now. It removes a bit of the headache of having to implement the prepare function for each subclass.

**Comment by bendmorris** [comment]
That seems reasonable to me. This looks good to me, I'd like to test it out on some real projects later tonight.

**Comment by MattTuttle** [comment]
I updated the code so that it's a lot easier to use.

```
var shader = new Shader(vertexCode, fragmentCode);
shader.position.name = "aPosition";
shader.texCoord.name = "aTexCoord";
shader.color.name = "aColor";
graphics.shader = shader;
```

**Comment by bendmorris** [comment]
Nice! This is very easy to use.

I imagine a common case will be extending TextureShader and using a custom fragment shader. What do you think about making TextureShader easy to extend? Maybe something like:

```
    function new(?vertex:String, ?fragment:String)
    {
        if (vertex == null) vertex = VERTEX_SHADER;
        if (fragment == null) fragment = FRAGMENT_SHADER;
        super(vertex, fragment);
```

Then if you're using the same names as TextureShader, the simple case is a one liner:

```
graphic.shader = new TextureShader(null, Assets.getText("shaders/my_shader.frag"));
```

**Comment by MattTuttle** [comment]
I like that idea. TextureShader and ColorShader can now be created with a different fragment shader.

```
graphic.shader = new TextureShader(Assets.getText("shaders/my_shader.frag"));
```

I figure if someone wants to replace the vertex shader they will do so with the default shader class.

**Pull request merged**